commit 3e0d14912e5560583800ca442e21458a1e2ba5f7
Author: Tim Serong <tserong@suse.com>
Date:   Sat May 13 00:16:14 2017 +1000

    mgr: load modules in separate python sub-interpreters
    
    This provides a reasonable amount of isolation between mgr
    modules.  Notably, with this change, it's possible to have more
    than one mgr module use cherrypy without conflicts.
    
    Each MgrPyModule gets its own python sub-interpreter.  The logger,
    the ceph_state module and sys.path need to be set up separately
    for each sub-interpreter, so all that happens in MgrPyModule's
    constructor (previously this was done on the main interpreter
    in PyModules::init()).
    
    Each sub-interpreter has its own python thread state.  The main
    interpreter also has a python thread state, but that is almost
    unused except for during setup and teardown of the whole beast.
    
    On top of that, an additional python thread state is necessary
    for each OS thread that calls into python code (e.g. the serve()
    method).
    
    Some care needs to be taken to ensure that the right thread state
    is active at the right time; note how the call to handle_pyerror()
    in PyModules::init() had to be moved inside MgrPyModle::load().
    PyModules should be using the main thread state, and MgrPyModule
    should usually be using its sub-interpreter thread state, except
    for very early in its constructor (before the sub-interpreter has
    been created), and in the serve() method where another python
    thread state is created to map to the separate OS thread that is
    responsible for invoking this method.
    
    The ceph_state module (PyState.cc) naturally has no idea what
    context it's being run in, so uses PyThreadState_Get() when it
    needs to know the python thread state.
    
    Signed-off-by: Tim Serong <tserong@suse.com>

diff --git a/src/mgr/PyModules.cc b/src/mgr/PyModules.cc
index 8cc3f674c0b..5ba0b69b90e 100644
--- a/src/mgr/PyModules.cc
+++ b/src/mgr/PyModules.cc
@@ -30,40 +30,12 @@
 
 #define dout_context g_ceph_context
 #define dout_subsys ceph_subsys_mgr
-
 #undef dout_prefix
-#define dout_prefix *_dout << "mgr[py] "
+#define dout_prefix *_dout << "mgr " << __func__ << " "
 
 // definition for non-const static member
 std::string PyModules::config_prefix;
 
-namespace {
-  PyObject* log_write(PyObject*, PyObject* args) {
-    char* m = nullptr;
-    if (PyArg_ParseTuple(args, "s", &m)) {
-      auto len = strlen(m);
-      if (len && m[len-1] == '\n') {
-	m[len-1] = '\0';
-      }
-      dout(4) << m << dendl;
-    }
-    Py_RETURN_NONE;
-  }
-
-  PyObject* log_flush(PyObject*, PyObject*){
-    Py_RETURN_NONE;
-  }
-
-  static PyMethodDef log_methods[] = {
-    {"write", log_write, METH_VARARGS, "write stdout and stderr"},
-    {"flush", log_flush, METH_VARARGS, "flush"},
-    {nullptr, nullptr, 0, nullptr}
-  };
-}
-
-#undef dout_prefix
-#define dout_prefix *_dout << "mgr " << __func__ << " "
-
 // constructor/destructor implementations cannot be in .h,
 // because ServeThread is still an "incomplete" type there
 
@@ -374,35 +346,16 @@ int PyModules::init()
   Py_SetProgramName(const_cast<char*>(PYTHON_EXECUTABLE));
   Py_InitializeEx(0);
 
-  // Some python modules do not cope with an unpopulated argv, so lets
-  // fake one.  This step also picks up site-packages into sys.path.
-  const char *argv[] = {"ceph-mgr"};
-  PySys_SetArgv(1, (char**)argv);
-
-  if (g_conf->daemonize) {
-    auto py_logger = Py_InitModule("ceph_logger", log_methods);
-#if PY_MAJOR_VERSION >= 3
-    PySys_SetObject("stderr", py_logger);
-    PySys_SetObject("stdout", py_logger);
-#else
-    PySys_SetObject(const_cast<char*>("stderr"), py_logger);
-    PySys_SetObject(const_cast<char*>("stdout"), py_logger);
-#endif
+  // Let CPython know that we will be calling it back from other
+  // threads in future.
+  if (! PyEval_ThreadsInitialized()) {
+    PyEval_InitThreads();
   }
-  // Populate python namespace with callable hooks
-  Py_InitModule("ceph_state", CephStateMethods);
 
   // Configure sys.path to include mgr_module_path
   std::string sys_path = std::string(Py_GetPath()) + ":" + get_site_packages()
                          + ":" + g_conf->mgr_module_path;
   dout(10) << "Computed sys.path '" << sys_path << "'" << dendl;
-  PySys_SetPath((char*)(sys_path.c_str()));
-
-  // Let CPython know that we will be calling it back from other
-  // threads in future.
-  if (! PyEval_ThreadsInitialized()) {
-    PyEval_InitThreads();
-  }
 
   // Drop the GIL and remember the main thread state (current
   // thread state becomes NULL)
@@ -412,11 +365,11 @@ int PyModules::init()
   boost::tokenizer<> tok(g_conf->mgr_modules);
   for(const auto& module_name : tok) {
     dout(1) << "Loading python module '" << module_name << "'" << dendl;
-    auto mod = std::unique_ptr<MgrPyModule>(new MgrPyModule(module_name, pMainThreadState));
+    auto mod = std::unique_ptr<MgrPyModule>(new MgrPyModule(module_name, sys_path, pMainThreadState));
     int r = mod->load();
     if (r != 0) {
       // Don't use handle_pyerror() here; we don't have the GIL
-      // or a thread state (this is deliberate).
+      // or the right thread state (this is deliberate).
       derr << "Error loading module '" << module_name << "': "
         << cpp_strerror(r) << dendl;
       // Don't drop out here, load the other modules
@@ -424,7 +377,7 @@ int PyModules::init()
       // Success!
       modules[module_name] = std::move(mod);
     }
-  } 
+  }
 
   return 0;
 }
